### PR TITLE
docs(typings): #6458 use a more suitable return type for waitUntil

### LINKED ISSUE #6458: Stricter type definition for `waitUntil`
from #6412

> I also noticed waitUntil is hardcoded to return Promise<any> - Maybe this should return Promise<void> because either it waits or it throws an error, we shouldn't care about what it returns. If we do, then maybe a generic for the return value? This was throwing me off because I thought the any was related to the original issue. https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/waitUntil.ts#L71

### PR DESCRIPTION
**Pull request by iamkenos**
## Proposed changes

- fixes #6458 
- basically just define the return type 😬 

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update - typings

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by christian-bromann** [comment]
> Can i request for an example where it would return a value other than true please?

Sure, I think the comments are wrong then, I double checked:

```js
› browser.waitUntil(() => 'foobar')
'foobar'
› browser.waitUntil(() => 123)
123
```

**Comment by iamkenos** [comment]
> > Can i request for an example where it would return a value other than true please?
> 
> Sure, I think the comments are wrong then, I double checked:
> 
> ```js
> › browser.waitUntil(() => 'foobar')
> 'foobar'
> › browser.waitUntil(() => 123)
> 123
> ```

i see, thanks! 

this is actually what i meant by - `...the check for that is already imposed by the argument type.`

image

one thing we can do is change the arg type to truthy instead of boolean, then combine it with generics to use the same arg type as the return type. it's [not gonna be straightforward](https://stackoverflow.com/questions/51390276/type-for-non-falsy-aka-truthy) but it will be more accurate.

personally, i've mixed feelings about this...thoughts?

File: packages/webdriverio/src/commands/browser/waitUntil.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ export default function waitUntil(
         interval = this.options.waitforInterval,
         timeoutMsg
     }: Partial<WaitUntilOptions> = {}
-) {
+): true | void {
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/waitUntil.ts:53]:
```suggestion
): Promise<true | void> {
```

**Comment by christian-bromann** [comment]
> one thing we can do is change the arg type to truthy instead of boolean, then combine it with generics to use the same arg type as the return type. it's [not gonna be straightforward](https://stackoverflow.com/questions/51390276/type-for-non-falsy-aka-truthy) but it will be more accurate.

That would be my preference given that this is how it actually works, regardless of the types. Also I can see that user would like to use the value returned by `waitUntil`

**Comment by iamkenos** [comment]
> > one thing we can do is change the arg type to truthy instead of boolean, then combine it with generics to use the same arg type as the return type. it's [not gonna be straightforward](https://stackoverflow.com/questions/51390276/type-for-non-falsy-aka-truthy) but it will be more accurate.
> 
> That would be my preference given that this is how it actually works, regardless of the types. Also I can see that user would like to use the value returned by `waitUntil`

Noted, thanks! Lemme see what i can do!

**Comment by iamkenos** [comment]
> > > one thing we can do is change the arg type to truthy instead of boolean, then combine it with generics to use the same arg type as the return type. it's [not gonna be straightforward](https://stackoverflow.com/questions/51390276/type-for-non-falsy-aka-truthy) but it will be more accurate.
> > 
> > 
> > That would be my preference given that this is how it actually works, regardless of the types. Also I can see that user would like to use the value returned by `waitUntil`
> 
> Noted, thanks! Lemme see what i can do!

hi @christian-bromann 

i tried [this](https://stackoverflow.com/a/51390763/2285470) approach and bumped into the limitation that was highlighted -- `Note how it won't allow anything which it thinks might be falsy...`

samples: 
```
browser.waitUntil(() => '') // type error, expected
browser.waitUntil(() => 123) // ok
browser.waitUntil(() => 'false') //ok
browser.waitUntil(() => true) // ok
browser.waitUntil(() => false) // type error, expected
browser.waitUntil(() => el.isClickable()) // type error, should be allowed but might be falsy so nope
```

for now i kept the changes minimal, as per last requests. thoughts?

**Review by christian-bromann — APPROVED** [review]
LGTM 👍  thank you so much!

**Pull request merged**